File: python/ctci/stacks_and_queues/stacks.py
# python implementation of a stack with a linked list (ADT)
class Stack(object):

    # Note that this should be a private object not used by app
    # Internal Node object for stacks
    class Node(object):
        def __init__(self, data=None, next=None):
            self.data = data
            self.next = next

    def __init__(self):
        self.head = None

    def push(self, item):
        cur_node = self.Node(item, self.head)
        self.head = cur_node

    def pop(self):
        if self.head is None:
            raise Exception('Empty Stack')
        ret_val = self.head.data
        self.head = self.head.next
        return ret_val

    def peek(self):
        if self.head is None:
            raise Exception('Empty Stack')
        return self.head.data

    def isEmpty(self):
        if self.head is None:
            return True
        return False

if __name__ == '__main__':
    st = Stack()
    st.push(1)
    print(st.peek())
    st.push(2)
    st.push(3)
    print(st.peek())
    print(st.isEmpty())
    print(st.pop())
    print(st.pop())
    st.pop()
    print(st.isEmpty())

File: cpp/stacks_queues/stacks.cc
#include <iostream>

using namespace std;

// Implement a stack using a linked list
class Stack {

  private:
    // Node class definition
    class Node {
      public:
        int data;
        Node* next;

        Node(int data, Node* next) {
          this->data = data;
          this->next = next;
        }
    };

  Node* head;

  public:
    // Default Constructor
    Stack() {
      this->head = NULL;
    }

    void push(int item) {
      Node* new_node = new Node(item, this->head);
      this->head = new_node;
    }

    int pop() {
      if (head == NULL) {
        throw "Stack Empty!";
      }
      Node* cur_head = this->head;
      int ret_val = cur_head->data;
      this->head = cur_head->next;
      free(cur_head);
      return ret_val;
    }

    int peek() {
      if (head == NULL) {
        throw "Stack Empty!";
      }
      return (this->head)->data;
    }

    bool isEmpty() {
      return (head == NULL);
    }
};

int main() {
  Stack st;
  st.push(1);
  cout << (st.peek()) << endl;
  st.push(2);
  cout << (st.peek()) << endl;
  st.push(3);
  cout << (st.peek()) << endl;
  cout << (st.isEmpty()) << endl;

  cout << (st.pop()) << endl;
  cout << (st.pop()) << endl;
  cout << (st.pop()) << endl;
  cout << (st.isEmpty()) << endl;
}